test-case/test-mic-privacy.sh

```diff
@@ -0,0 +1,180 @@
+
+OPT_NAME['r']='rate'        OPT_DESC['r']='sample rate'
+OPT_HAS_ARG['r']=1          OPT_VAL['r']=48000
+
+OPT_NAME['R']='relay'       OPT_DESC['R']='name of usbrelay switch, default value is HURTM_1'
+OPT_HAS_ARG['R']=1          OPT_VAL['R']='HURTM_1'
+
+func_opt_parse_option "$@"
+
+pcm_p=${OPT_VAL['p']}
+pcm_c=${OPT_VAL['c']}
+channel_c=${OPT_VAL['C']}
+channel_p=${OPT_VAL['N']}
+rate=${OPT_VAL['r']}
+relay=${OPT_VAL['R']}
+
+dlogi "Params: pcm_p=$pcm_p, pcm_c=$pcm_c, channel_c=$channel_c, channel_p=$channel_p, rate=$rate, LOG_ROOT=$LOG_ROOT"
+
+__upload_wav_files()
+{
+    # upload the alsabat wav file
+    for file in $ALSABAT_WAV_FILES; do
+        # alsabat has a bug where it creates an empty record in playback mode
+        if test -s "$file"; then
+            cp -v "$file" "$LOG_ROOT/"
+        fi
+    done
+}
+
+check_playback_capture()
+{
+    # check if capture and playback work
+    dlogc "alsabat -P$pcm_p -C$pcm_c -c 2 -r $rate"
+    alsabat -P"$pcm_p" -C"$pcm_c" -c 2 -r "$rate" || {
+        # upload failed wav file
+        __upload_wav_files
+        dle "alsabat failed"
+    }
+}
+
+main()
+{
+    setup_kernel_check_point
+
+    start_test
+
+    logger_disabled || func_lib_start_log_collect
+
+    if [ "$pcm_p" = "" ]||[ "$pcm_c" = "" ]; then
+        skip_test "No playback or capture PCM is specified. Skip the alsabat test."
+    fi
+
+    # check if usbrelay tool is installed
+    command -v usbrelay || {
+        skip_test "usbrelay command not found. Please install usbrelay to control the mic privacy switch."
+    }
+
+    check_locale_for_alsabat
+
+    # reset sof volume to 0dB
+    reset_sof_volume
+
+    # If MODEL is defined, set proper gain for the platform
+    if [ -z "$MODEL" ]; then
+        # treat as warning only
+        dlogw "NO MODEL is defined. Please define MODEL to run alsa_settings/MODEL.sh"
+    else
+        dlogi "apply alsa settings for alsa_settings/MODEL.sh"
+        set_alsa_settings "$MODEL"
+    fi
+
+    dlogi "Turn off the mic privacy switch"
+    usbrelay_switch "$relay" 0
+
+    # check the PCMs before mic privacy test
+    dlogi "check the PCMs before mic privacy test"
+    aplay   -Dplug"$pcm_p" -d 1 /dev/zero -q || die "Failed to play on PCM: $pcm_p"
+    arecord -Dplug"$pcm_c" -d 1 /dev/null -q || die "Failed to capture on PCM: $pcm_c"
+
+    # Select the first card
+    first_card_name=$(aplay -l | awk '/^card ([0-9]+)/ {print $3; exit}')
+    # dump amixer contents always.
+    # Good case amixer settings is for reference, bad case for debugging.
+    amixer -c "${first_card_name}" contents > "$LOG_ROOT"/amixer_settings.txt
+
+    check_playback_capture
+
+    dlogi "Turn on the mic privacy switch"
+    usbrelay_switch "$relay" 1
+
+    alsabat_output=$(mktemp)
+    dlogc "alsabat -P$pcm_p -C$pcm_c -c 2 -r $rate"
+    alsabat -P"$pcm_p" -C"$pcm_c" -c 2 -r "$rate" >"$alsabat_output" 2>&1
+    alsabat_status=$?
+
+    if [ $alsabat_status -ne 0 ]; then
```

Comment:
Prefer the "early return" pattern
```suggestion
    alsabat -P"$pcm_p" -C"$pcm_c" -c 2 -r "$rate" >"$alsabat_output" 2>&1 || {
            dloge "alsabat passed unexpectedly, upload the wav files."
        __upload_wav_files
        die "MIC privacy doesn't work. alsabat output: $(cat "$alsabat_output")"
    }
```